src/Equations/Fluid/LCFSLossRateTerm.cpp

```diff
@@ -0,0 +1,166 @@
+ */
+
+#include "DREAM/DREAMException.hpp"
+#include "DREAM/Equations/Fluid/LCFSLossRateTerm.hpp"
+
+using namespace DREAM;
+
+/**
+ * Constructor.
+ */
+LCFSLossRateTerm::LCFSLossRateTerm(FVM::Grid *grid, FVM::UnknownQuantityHandler *unknowns, FVM::Grid *operandGrid, real_t sf, real_t* t_loss, len_t userGivenPsiEdge_t0, real_t PsiEdge_t0) : 
+    RunawaySourceTerm(grid, unknowns), FVM::DiagonalComplexTerm(grid, unknowns, operandGrid), unknowns(unknowns), scaleFactor(sf), t_loss(t_loss), 
+    userGivenPsiEdge_t0(userGivenPsiEdge_t0), psi_edge_t0(PsiEdge_t0), 
+    id_psi(unknowns->GetUnknownID(OptionConstants::UQTY_POL_FLUX)) {
+
+    SetName("LCFSLossRateTerm");
+    this->GridRebuilt();
+}
+
+/**
+ * Destructor
+ */
+LCFSLossRateTerm::~LCFSLossRateTerm(){
+    Deallocate();
+}           
+
+
+/**
+ * Called after the grid is rebuilt; (re)allocates memory
+ * for all quantities
+ */
+bool LCFSLossRateTerm::GridRebuilt(){
+
+    this->DiagonalComplexTerm::GridRebuilt();
+    Deallocate();
+    this->GammaLoss = new real_t[nr];
+    
+    return true;
+}
+
+
+/**
+* Set weights for term to be multiplied with unknown (n_RE). 
+*/
+void LCFSLossRateTerm::SetGammaLoss(){
+
+    for (len_t ir = 0; ir < nr; ir++){ // Empty GammaLoss
+        this->GammaLoss[ir] = 0;
+    }
+    
+    FindRadiusOfLCFS(); // Find ir_LCFS
+    
+    for (len_t ir = 0; ir < nr; ir++){
+        this->GammaLoss[ir] = -StepFunction(ir) / this->t_loss[ir]; // Change GammaLoss from 0 if ir>ir_LCFS
+    }
+}
+
+
+
+/**
+* Find current radial point of last closed flux surface
+* and find psi_edge_t0 if not provided in input
+*/
+void LCFSLossRateTerm::FindRadiusOfLCFS(){
+    
+    if(this->userGivenPsiEdge_t0 == 0) {
+        this->psi_edge_t0 = unknowns->GetUnknownInitialData(id_psi)[nr-1];
+    }
+    
+    // Find if the sign of PsiDiff() should be +1 or -1
+    if(this->signFixed == false) { 
+        real_t psi_mid_init = unknowns->GetUnknownInitialData(id_psi)[0]; 
+        real_t psi_edge_init = unknowns->GetUnknownInitialData(id_psi)[nr-1];
+        if(psi_mid_init-psi_edge_init > 0) { this->sign = 1; }
+        this->signFixed = true;
+    }
+    
+    bool exists = false;
+    for (len_t ir = 0; ir < nr; ir++){
+        
+        if(this->sign * PsiDiff(ir) >= 0){
+            this->ir_LCFS = ir; // Last ir to pass critera is ir_LCFS
+            exists = true;
+        }
+    }
+    if ( exists == false ) { this->ir_LCFS = -1; } // No flux surface is closed
+}
+
+
+
+/**
+* Difference between the poloidal flux at radial point ir and at the edge at t=0
+*/
+real_t LCFSLossRateTerm::PsiDiff(len_t ir){
+    real_t psi = unknowns->GetUnknownDataPrevious(id_psi)[ir]; // Previous data to skip dependence of psi in Jacobian
+    return psi - this->psi_edge_t0;
+}
+
+
+
+/**
+* Currently a step function at ir_LCFS, could be changed to continuous
+*/
+real_t LCFSLossRateTerm::StepFunction(len_t ir){
+    real_t x;
```

Comment:
A very picky comment but you don't need the variable x here, you can just write the content of this function as

```
    if (ir > this->ir_LCFS) { 
        return = 1; 
    } else { 
        return = 0; 
    }
```